presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcStorageManager.java

```diff
@@ -393,55 +386,20 @@ private static OrcFileInfo rewriteFile(File input, File output, BitSet rowsToDel
         }
     }
 
-    private List<ColumnInfo> getColumnInfo(OrcReader reader)
+    private List<ColumnInfo> getColumnInfo(OrcReader reader, List<Type> columnTypes)
     {
-        // TODO: These should be stored as proper metadata.
-        // XXX: Relying on ORC types will not work when more Presto types are supported.
-
         List<String> names = reader.getColumnNames();
-        Type rowType = getType(reader.getFooter().getTypes(), 0);
-        if (names.size() != rowType.getTypeParameters().size()) {
+        if (names.size() != columnTypes.size()) {
             throw new PrestoException(RAPTOR_ERROR, "Column names and types do not match");
         }
 
         ImmutableList.Builder<ColumnInfo> list = ImmutableList.builder();
         for (int i = 0; i < names.size(); i++) {
-            list.add(new ColumnInfo(Long.parseLong(names.get(i)), rowType.getTypeParameters().get(i)));
+            list.add(new ColumnInfo(Long.parseLong(names.get(i)), columnTypes.get(i)));
```

Comment:
Are we sure that the column order returned from metadata will be always the same as a column order returned by `reader.getColumnNames`. Maybe it is better to pass list of `ColumnHandle` instead of list of `Type`, and match the column by its name?